spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala

```diff
@@ -0,0 +1,366 @@
+    gradInput
+  }
+}
+
+// Return postition encoding with input shift right
+private[nn] class PositionEncodeWithShift[T: ClassTag](implicit ev: TensorNumeric[T])
+  extends TensorModule[T] {
+
+  @transient private var rangeBuffer : Tensor[T] = null
+  @transient private var timeBuffer : Tensor[T] = null
+
+  // input a Tensor with shape [batch, length, channels]
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    TransformerOperation.shiftRight3D(input, output)
+    val length = output.size(2)
+    val channel = output.size(3)
+
+    if (rangeBuffer == null) {
+      rangeBuffer = Tensor[T]()
+      TransformerOperation.initRangeTensor(length, rangeBuffer)
+    }
+    if (timeBuffer == null) {
+      timeBuffer = Tensor[T]().resize(length, channel)
+      TransformerOperation.addTimingSignal1D(length, channel,
+        rangeBuffer = rangeBuffer, timeBuffer = timeBuffer)
+    }
+    val batchSize = input.size(1)
+    var i = 1
+    while (i <= batchSize) {
+      output.select(1, i).add(timeBuffer)
+      i += 1
+    }
+    return output
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    if (gradInput == null) gradInput = Tensor[T]()
+    gradInput.resizeAs(gradOutput).zero()
+    val size = gradOutput.size(2)
+    var i = 1
+    while (i < size) {
+      gradInput.select(2, i).copy(gradOutput.select(2, i + 1))
+      i += 1
+    }
+    gradInput
+  }
+}
+
+/**
+ * Calculate bias tensor from padding values in tensor.
+ * Bias tensor that is added to the pre-softmax multi-headed attention logits,
+ * which has shape [batch_size, num_heads, length, length]. The tensor is zero at
+ * non-padding locations, and -1e9 (negative infinity) at padding locations.
+ * @param ev$1
+ * @param ev
+ * @tparam T The numeric type in this module parameters
+ */
+private[nn] class PaddingMask[T: ClassTag](implicit ev: TensorNumeric[T])
+  extends TensorModule[T] {
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    output.resizeAs(input).copy(input)
+    output = TransformerOperation.getPaddingBias(output)
+    output
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    gradInput.resizeAs(input).zero()
+    gradInput
+  }
+}
+
+// This mask is to hide both <pad> and future words. Used in decode
+private[nn] class SelfAttentionMask[T: ClassTag](implicit ev: TensorNumeric[T])
+  extends TensorModule[T] {
+  /**
+   * Create an bias tensor to be added to attention logits.
+   * Returns tensor with shape (1, 1, length, length)
+   * @param length
+   * @tparam T
+   * @return
+   */
+  private def attentionBiasLowerTriangle[T: ClassTag](
+    length: Int, output: Tensor[T])(implicit ev: TensorNumeric[T]): Tensor[T] = {
+    val arr = output.storage().array()
+    for (i <- 0 to (length - 1)) {
+      var j = length - 1
+      while (j > i) {
+        // reminder: here not 1
+        arr(i * length + j) = ev.fromType(-1e9)
```

Comment:
Any more efficient way for setting value with constant

Reply:
ok,changed